File: app/src/main/java/com/assignment/presentation/actvities/PhotoViewerActivity.java
package com.assignment.presentation.actvities;

import android.os.Bundle;
import androidx.viewpager.widget.ViewPager;
import androidx.appcompat.widget.Toolbar;
import android.widget.TextView;

import com.assignment.R;
import com.assignment.common.Constants;
import com.assignment.data.SearchPhotoDataModel;
import com.assignment.data.StatusData;
import com.assignment.presentation.BaseActivity;
import com.assignment.presentation.adapters.PhotoViewerAdapter;
import com.assignment.presentation.viewmodels.common.BaseViewModel;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import dagger.android.AndroidInjection;

public class PhotoViewerActivity extends BaseActivity {
    private Toolbar mToolbar;
    private TextView mPhotoCountTextView;
    private ViewPager mImageViewPager;
    private List<SearchPhotoDataModel.PhotoItem> mPhotos;
    private int mSelectedPosition, mCurrentPosition;

    @Inject
    Gson gson;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_full_screen;
    }

    @Override
    protected void initializeViews(Bundle bundle) {
        AndroidInjection.inject(this);
        mToolbar = findViewById(R.id.toolbar);
        setSupportActionBar(mToolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(false);
            getSupportActionBar().setDisplayShowHomeEnabled(false);
            getSupportActionBar().setTitle("");
        }

        mPhotoCountTextView = findViewById(R.id.tv_photo_count);
        mImageViewPager = findViewById(R.id.vp_image);
        getDataFromIntent();

        if (mPhotos != null && mPhotos.size() != 0) {
            String countOnTextView = mSelectedPosition+1 + "/" + mPhotos.size();
            mPhotoCountTextView.setText(countOnTextView);
            PhotoViewerAdapter photoViewerAdapter = new PhotoViewerAdapter(this, mPhotos);
            mImageViewPager.setAdapter(photoViewerAdapter);

            mImageViewPager.setCurrentItem(mSelectedPosition);

            mImageViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
                @Override
                public void onPageScrolled(int i, float v, int i1) {

                }

                @Override
                public void onPageSelected(int i) {
                    mCurrentPosition = i+1;
                    String countOnTextView = i + "/" + mPhotos.size();
                    mPhotoCountTextView.setText(countOnTextView);
                }

                @Override
                public void onPageScrollStateChanged(int i) {

                }
            });
        }


    }

    private void getDataFromIntent() {
        Type type = new TypeToken<ArrayList<SearchPhotoDataModel.PhotoItem>>() {
        }.getType();
        mPhotos = gson.fromJson(getIntent().getStringExtra(Constants.Extras.PHOTOS), type);
        mSelectedPosition = getIntent().getIntExtra(Constants.Extras.SELECTED_POSITION, 0);
        mCurrentPosition = mSelectedPosition;
    }

    @Override
    protected BaseViewModel initViewModel() {
        return null;
    }

    @Override
    public void onBackPressed() {

        if (mSelectedPosition == mCurrentPosition){
            super.onBackPressed();
        }else {
            finish();
        }

    }

    @Override
    protected void handleViewModelUpdatesOnSuccess(StatusData status) {

    }

    @Override
    protected void handleLiveData() {

    }

    @Override
    protected void handleViewModelUpdatesOnFailure(StatusData status, Throwable throwable) {

    }
}

File: app/src/main/java/com/assignment/presentation/actvities/PhotoViewerActivity.kt
package com.assignment.presentation.actvities

import android.os.Bundle
import android.util.Log
import androidx.lifecycle.ViewModelProviders
import androidx.viewpager.widget.ViewPager
import com.assignment.BR
import com.assignment.R
import com.assignment.common.Constants
import com.assignment.databinding.ActivityFullScreenBinding
import com.assignment.presentation.BaseActivity
import com.assignment.presentation.adapters.PhotoViewerAdapter
import com.assignment.presentation.helpers.InternetStatusImpl
import com.assignment.presentation.viewmodels.PhotoViewerModel
import com.assignment.presentation.viewmodels.ViewModelFactory
import com.assignment.presentation.viewmodels.common.SingleLiveEvent
import com.assignment.usecase.Intractors
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.model.PhotoItem
import dagger.android.AndroidInjection
import kotlinx.android.synthetic.main.activity_full_screen.*
import java.util.*
import javax.inject.Inject

class PhotoViewerActivity : BaseActivity<ActivityFullScreenBinding, PhotoViewerModel>() {

    private var mPhotos: List<PhotoItem>? = null
    private var mSelectedPosition: Int = 0
    private var mCurrentPosition: Int = 0
    @Inject
    lateinit var gson: Gson
    @Inject
    lateinit var internetStatusImpl: InternetStatusImpl
    @Inject
    lateinit var intractors: Intractors

    override fun getLayoutId(): Int {
        return R.layout.activity_full_screen
    }

    override fun initializeViews(bundle: Bundle?) {
        AndroidInjection.inject(this)
        setSupportActionBar(toolbar)
        supportActionBar?.let {
            it.setDisplayHomeAsUpEnabled(false)
            it.setDisplayShowHomeEnabled(false)
            it.title = ""
        }
    }

    override fun onResume() {
        super.onResume()
        getDataFromIntent()

        mPhotos?.let {
            viewModel?.currentPage?.set((mSelectedPosition + 1).toString() + "/" + it.size)
            vp_image?.adapter = PhotoViewerAdapter(this, it)
            vp_image?.currentItem = mSelectedPosition
            vp_image?.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                override fun onPageScrolled(i: Int, v: Float, i1: Int) {
                    Log.v("PhotoViewActivity", "onPageScrolled")
                }

                override fun onPageSelected(i: Int) {
                    mCurrentPosition = i + 1
                    viewModel?.currentPage?.set(mCurrentPosition.toString() + "/" + it.size)
                }

                override fun onPageScrollStateChanged(i: Int) {
                    Log.v("PhotoViewActivity", "onPageScrollStateChanged")
                }
            })
        }

    }
    private fun getDataFromIntent() {
        val type = object : TypeToken<ArrayList<PhotoItem>>() {

        }.type
        mPhotos = gson.fromJson<List<PhotoItem>>(intent.getStringExtra(Constants.Extras.PHOTOS), type)
        mSelectedPosition = intent.getIntExtra(Constants.Extras.SELECTED_POSITION, 0)
        mCurrentPosition = mSelectedPosition
    }

    override fun initViewModel(): PhotoViewerModel {
        val viewModelFactory = ViewModelFactory(this.application, intractors
                , SingleLiveEvent(), internetStatusImpl)
        return ViewModelProviders.of(this, viewModelFactory).get(PhotoViewerModel::class.java)
    }

    override fun onBackPressed() {
        if (mSelectedPosition == mCurrentPosition) {
            super.onBackPressed()
        } else {
            finish()
        }

    }

    override fun getBindingVariable(): Int {
        return BR.viewmodel
    }

}
